<img style="float:center" src="https://upload.wikimedia.org/wikipedia/commons/d/db/Logo_ITESO_normal.jpg" width="150px"/>
<div align="center">

____
## SIMULACIÓN MATEMÁTICA
### PROYECTO MÓDULO II:
### Cálculo de la carga eléctrica encerrada en un sistema por medio de integración Montecarlo

____
<div align="left">
    
__Integrantes__
+ Adriana Del Río Ramírez
    - Estudiante de Ing. en nanotecnología
        - nt717047 
+ Iván Madrazo Álvarez
    - Estudiante de Ing. en nanotecnología 
        - nt710381

+ Rodrigo Ayala Padilla
    + Estudiante de Ing. en nanotecnología
        + nt716733

## INTRODUCCIÓN


Uno de los problemas más comunes de teoría electromagnética, consta de la obtención del campo eléctrico encerrado por dos esferas concéntricas, una de menor tamaño que la otra como se muestra en la siguiente imagen:

![Image](https://dademuchconnection.files.wordpress.com/2018/08/null27.png?w=364&h=296)

La esfera más pequeña describe a un cascarón es decir una esfera vacía de radio a, sin embargo, para poder realizar la integración Montecarlo asignaremos un valor real de $1 cm$, mientras que la esfera mayor que rodea al cascarón posee un radio b de $2 cm$ y se encentra rellena de un material aislante dieléctrico, para hacer más fácil el cálculo, el sistema se dispondrá en el vacío. 

El problema plantea lo siguiente; Calcular el campo eléctrico en los puntos en que este no sea cero para el sistema que presenta una polarización $P0 = 3 cm^2$. Para solucionar el problema, primero es importante conocer ciertos conceptos que nos serán de ayuda descritos a continuación.

#### Ley de Gauss
>La ley de Gauss fue descrita por el físico, astrónomo y matemático, Johann Carl Friedrich Gauss y establece que el flujo eléctrico, que atraviesa cualquier superficie, es proporcional a la carga que dicha superficie encierra. Sin embargo, la principal limitante de esta ley es que el caso que se analice debe poseer gran simetría, dado que se debe poder “encerrar” la zona de interés por una superficie ficticia, conocida como superficie gaussiana y dicha superficie debe ser una superficie conocida para que este método tenga relevancia.
#### Dieléctricos
>Al estar sometido a un campo eléctrico, un cuerpo puede cambiar sus propiedades de diferentes maneras. Esto se debe a que los cuerpos están compuestos por partículas cargadas, los cuales interactúan con dicho campo eléctrico.   
>Los tipos más comunes de material o sustancias, según sus propiedades eléctrica, se pueden clasificar en conductores y dieléctricos (no conductores).
>La principal característica que distingue a un material conductor de uno dieléctrico, es que a diferencia de los conductores, el dieléctrico no permite el libre movimiento de partículas cargadas, los cual hace que no pueda fluir una corriente eléctrica en éste. A pesar de no haber movimiento en las partículas cargadas, los dieléctricos si presentan interacción con un campo eléctrico externo, lo cual permite que ocurran fenóenos tales como el dipolo inducido.

(Redondo Quintera, F. & Redondo Melchor, R., 2007)

#### Dipolo inducido
>Si bien, un átomo en su totalidad es eléctricamente neutro y mediante tal afirmación se podría llegar a la conclusión de que el campo eléctrico no le debería de afectar, tanto su núcleo, como la nube electrónica que lo rodea se tienen regiones eléctricamente cargadas, las cuales si se verán afectadas por algún campo eléctrico externo. Si el campo que se aplica no es demasiado fuerte, se establece un equilibrio rápidamente, de tal forma que si el centro de la nube electrónica no coincide con el núcleo, hace que estas cargas se atraen mutuamente, manteniendo al átomo unido.   
>Ahora, si tomamos en cuenta que el campo eléctrico que se aplique sobre un átomo cualquiera, generara que una fuerza que intenta separar tanto la nube electrónica del átomo de su núcleo, mientras que la atracción mutua los mantiene unidos, esta combinación de fuerzas llega a un balance llamado polarización, ya que tanto la carga positiva como la negativa están ligeramente desplazadas de su punto de equilibrio natural. Esto genera un ligero momento dipolar (lo que genera un dipolo) en el átomo y puesto que no es algo propio del átomo en sí, sino que esta dado gracias a la acción e un campo externo, se le nombra dipolo inducido.

(Becerra, A., 2007)

#### Polarización
>El desplaza miento de las partículas cargadas genera un momento dipolar. ​
El momento dipolar generado en cada unidad del material, se conoce como densidad de polarización (Que es un campo vectorial).​

(Electrostática en presencia de dieléctricos, 2011)

## OBJETIVOS

__`Objetivo general`__
>Hacer uso de la integración Montecarlo para obtener la carga encerrada de un cascarón dieléctrico encerrado por otro de mayor tamaño, de esta forma obtener el campo eléctrico dentro en una esfera doble polarizada de cierta forma.

__`Objetivos específicos`__
+ Utilizar el método de integración Montecarlo visto en clase, con el módulo random de la libreria numpy para el cálculo de integrales, en este caso siendo una integral triple.
+ Utilizar la librería pandas para el acomodo de los ajustes realizados por la integración Montecarlo para observar cómo evoluciona la aproximación de la integración con el número de términos empleados, con respecto al resultado real de la integración.

## MODELO

Dada la simetría del sistema, se puede suponer que el potencial eléctrico depende exclusivamente de la distancia r al centro del sistema y el campo eléctrico es, por lo tanto, central. Fácilmente se puede partir de la aplicación de la ley de Gauss para resolver el sistema. En este caso, el flujo del campo eléctrico a través de la superficie esférica concéntrica se describe de la siguiente manera:

$$\oint { \vec { E } \cdot d\vec { a } =\frac { { Q }_{ e } }{ {\varepsilon}_{ 0 } }} \rightarrow \vec{E}=\frac{Q_e}{4\pi r^2 \varepsilon_0}  \hat{r} = \overrightarrow{E} = \frac{-18 \pi}{4\pi 1.5^2 8.85x10^{-12}}  \hat{r} = \frac{-18 \pi}{4\pi 1.5^2 8.85x10^{-12}}  \hat{r} $$


$$\overrightarrow{E} = -226028274091.20285[𝑁/𝐶]  $$ 

Donde:
- ${ Q }_{ e }$ es la carga encerrada por la superficie gaussiana
- $\varepsilon_0$ es la permitibilidad eléctrica del vacío
- $r$ es el radio de la superficie gaussiana

Donde todos los parámetros son conocidos, a excepción de la carga encerrada por la superficie gaussiana. 
Al tratarse de materiales dieléctricos, consideramos una polarización inducida en el material, por lo que los cascarones esféricos se encuentran polarizados, es decir existe carga no nula. 
Ya que la carga de polarización (o cargas ligadas),  se deben a un volumen polarizado, ésta puede transformarse en suma de dos integrales:

$$Q_e= \int_{\tau}^{} \rho_b d \tau  +\oint_{S}^{}\sigma_{ab} d S $$

Donde:
- $\tau$ es el volúmen que ocupa el dieléctrico polarizado, y $d \tau$ es el diferencial de volumen
- $S$ es la superficie cerrada que limita a $\tau$, y $d S$ es el diferencial de superficie
- $\rho_b$ es la densidad volumétrica de carga ligada
- $\sigma_{p}$ es la densidad superficial de carga ligada

La ventaja de esta transformación, es que este tipo de integrales corresponden a la carga debida a la distribución de carga eléctrica.
En términos físicos, esto podría entenderse como que una polarización  es equivalente a una superposición de dos distribuciones de carga, una volumétrica y otra superficial.
Una reformulación en las ecuaciones de Maxwell, puede permitir expresarlas en función del vector de densidad de polarización, donde:

$$\sigma_{p}=\overrightarrow{P}\cdot\hat{n}$$

$$\rho_b=-\triangledown\cdot\overrightarrow{P}$$

Aquí $\overrightarrow{P}$ es el vector desnidad de polarización
El vector densidad de polarización se compone por la dirección radial del momento dipolar generado y por la densidad de polarización:

$$\overrightarrow { P } =\frac{{ P }_{ 0 }\widehat { r }}{r} $$

De modo que la densidad volumétrica de carga de polarización tiene la expresión general

$$\rho_b=-\triangledown\cdot\frac{{ P }_{ 0 }\widehat { r }}{r}$$

Haciendo uso de la siguiente identidad sobre la divergencia de una función en coordenadas esféricas: 

$$-\triangledown\cdot\overrightarrow{F}= -(\frac{1}{r^2}\frac{\partial{}}{\partial{r}}(r^2 F_r)) + \frac{1}{r \sin(\theta)}\frac{\partial{}}{\partial{\theta}}(\sin(\theta) F_\theta)) + \frac{1}{r \sin(\phi)}\frac{\partial{}}{\partial{\phi}}(\sin(\phi) F_\phi))$$

Y tomando en cuenta que sólo existe un vector radial sin componentes angulares, la densidad de volumétrica de carga ligada resulta:

$$-\triangledown\cdot\overrightarrow{P}= -(\frac{1}{r^2}\frac{\partial{}}{\partial{r}}(r^2 P_r))$$

$$-\triangledown\cdot\overrightarrow{P}= -(\frac{1}{r^2}\frac{\partial{}}{\partial{r}}(r^2 \frac{P_0}{r})=\rho_b$$

Mientras que la densidad superficial de carga de polarización, dado un volumen $\tau$, viene dado por la expresión:

$$\sigma_p=\frac{{ P }_{ 0 }\widehat { r }}{r} \cdot \hat{n}$$

Esta expresión se puede generalizar al caso de que se tenga una interfaz entre dos dieléctricos con polarización P1 y P2, en cuyo caso la densidad superficial será la suma de las correspondientes a cada dieléctrico. 

$$\sigma_p=\sigma_a + \sigma_b = \overrightarrow{P_a}\cdot\hat{n} + \overrightarrow{P_b}\cdot\hat{n}$$

Por último, las densidades superficiales de carga ligada se descomponen en el vector de polarización producto punto con el vector unitario normal a la superficie. Así, las dos densidades de carga superficiales correspondientes a cada cascarón resultan:

$$\sigma_a=\frac{P_0}{r_a} \hat{r} \cdot \hat{-r}$$

$$\sigma_b=\frac{P_0}{r_b} \hat{r} \cdot \hat{r}$$

Volviendo a la expresión de carga ligada, se analizará el caso en donde $a<r<b$, el cual queda como sigue:

$$Q_e= \int_{0}^{2\pi} \int_{0}^{\pi} \int_{1}^{1.5} -\frac{3}{r^2}r^2\sin{\theta} d r d \theta d \varphi +\int_{0}^{2\pi} \int_{0}^{\pi} -\frac{3}{1}1^2\sin{\theta} d \theta d \varphi$$


#### Nota:

Para el caso del analisis dentro de la esfera, se tiene que la carga que se encierra mediante la superfiecie gaussiana es inexistente (ya que la esfera es hueca). Por lo tanto, el campo dentro de la esfera es cero $$\overrightarrow{E} = 0$$


Mientras que si analizamos la parte externa de la esfera, se obtiene un modelo matemático similar al caso intermedio, con la excepción de que se debe de incluir una integral superficial de la parte externa, de tal modo que (por facilidad no se hará uso de limites establecidos ni de la polarización inicial):

$$Q_e= \int_{0}^{2\pi} \int_{0}^{\pi} \int_{a}^{b} -\frac{P_0}{r^2}r^2\sin{\theta} d r d \theta d \varphi +\int_{0}^{2\pi} \int_{0}^{\pi} -\frac{P_0}{a}a^2\sin{\theta} d \theta d \varphi + \int_{0}^{2\pi} \int_{0}^{\pi} -\frac{P_0}{b}b^2\sin{\theta} d \theta d \varphi$$

Las integrales dan como resultado lo siguiente:

$$ Q_e = 4\pi P_0 (a-b) + 4\pi P_0 b - 4\pi P_0 a = 4\pi P_0 (a-b) + 4\pi P_0 (b-a) = 0  $$

Lo que da como resultado $$\overrightarrow{E} = 0$$ 


Por lo que ambos resultados dan cero y el único punto donde es relevante calcular el campo eléctrico es en la parte intermedia del cascaron.

# Solución del problema

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline

In [2]:
#Definición de la integral volumétrica

def integral_volumen(r,theta):
    return ((-3*r**2)/(r**2))*np.sin(theta)

In [3]:
#Definición de la función montecarlo

def integracion_montecarlo(integral_volumen,a1,b1,a2,b2,a3,b3,N):
    Vr=(b1-a1)*(b2-a2)*(b3-a3)
    r=np.random.uniform(a1,b1,N)
    theta=np.random.uniform(a2,b2,N)
    return Vr*np.mean(integral_volumen(r,theta))

In [4]:
#Limites y creacion del data frame

a1,b1,a2,b2,a3,b3=1,1.5,0,np.pi,0,2*np.pi
N=np.logspace(1,7,7).astype(int)

solucion = pd.DataFrame(index=N,columns=['Aprox integral de volumen'])

for n in N:
    solucion.loc[n,'Aprox integral de volumen']= integracion_montecarlo(integral_volumen,a1,b1,a2,b2,a3,b3,n)
    
solucion

Aprox integral de volumen
10                        -20.8204
100                       -19.4981
1000                      -19.0408
10000                     -18.8465
100000                    -18.8248
1000000                   -18.8478
10000000                  -18.8526

In [5]:
#Definición de la integral de superficie

def integral_superficie(theta):
    return (-3/1)*(1)*np.sin(theta)

In [6]:
#Definición de la solucion de la integral superficial mediante montecarlo

def integracion_montecarlo2(integral_superficie,a1,b1,a2,b2,N):
    Vr=(b1-a1)*(b2-a2)
    theta=np.random.uniform(a1,b1,N.astype(int))
    return Vr*np.mean(integral_superficie(theta))

In [7]:
#Limites y data frame

a1,b1,a2,b2=0,np.pi,0,2*np.pi
N=np.logspace(1,7,7)

solucion2=pd.DataFrame(index=N,columns=['Aprox integral superficial'])

for n in N:
    solucion2.loc[n,'Aprox integral superficial']= integracion_montecarlo2(integral_superficie,a1,b1,a2,b2,n)
solucion2

Aprox integral superficial
10.0                         -36.4576
100.0                        -37.8103
1000.0                       -37.4773
10000.0                      -37.9337
100000.0                     -37.7154
1000000.0                    -37.6987
10000000.0                    -37.706

**Entre mayor sea la cantidad de terminos que se utilice en el método, mejor será el resultado. Es por esta razon que se utiliza la aproximación con 10 millones de valores.**

In [8]:
# Valor aproximado para la carga encerrada

carga_encerrada = solucion.iloc[6,0] + solucion2.iloc[6,0]
carga_encerrada

-56.55857016319426

In [9]:
#Campo electrico calculado mediante el valor aproximado de la carga encerrada

E=carga_encerrada/(4*np.pi*(1.5**2)*(8.85*(10**(-12))))
E

-226028274091.20285

Utilizando el valor obtenido para la carga encerrada, mediante el método montecarlo, se obtuvo el campo electrico para la region comprendida en el cascaron. Dando como resultado $-226028274091.20285[N/C]$

Para Poder calcular el error obtenido en la aproximación, es necesario crear una tabla con la suma de los valores para cada aproximación, para poder incluir esos valores en la resta con el valor real.


In [10]:
#Creacion de la tabla que contiene la suma de los valores aproximados para cada integral, usando x como cambio de fila.



tabla = pd.DataFrame(index = N, columns = ['Valor aproximado Qe'])

x = 0

for n in N:
    
    tabla.loc[n,'Valor aproximado Qe']= solucion.iloc[x,0] + solucion2.iloc[x,0]
    
    x  = x +  1
    
tabla




Valor aproximado Qe
10.0                   -57.278
100.0                 -57.3083
1000.0                -56.5181
10000.0               -56.7802
100000.0              -56.5401
1000000.0             -56.5466
10000000.0            -56.5586

El valor real obtenido mediante integración directa es de $-18\pi$. Por tanto solo falta restar este valor de de los valores caclulados por aproximación, para poder obtener el error.

In [11]:
#Error obtenido 

Valor_real = -18*np.pi

tabla2 = tabla

tabla2['error']=np.abs(Valor_real-tabla2['Valor aproximado Qe'])

tabla2

Valor aproximado Qe       error
10.0                   -57.278    0.729343
100.0                 -57.3083    0.759679
1000.0                -56.5181   0.0306145
10000.0               -56.7802    0.231579
100000.0              -56.5401  0.00854909
1000000.0             -56.5466  0.00209098
10000000.0            -56.5586   0.0099024

# conclusiones

Mediante el método de integracion Montecarlo, en conjunto con el modulo random de numpy, para la generacion de números aleatorios con distribución uniforme, variando la cantidad de numeros generados en potencias de 10, iniciando en 10 numeros aleatorios, hasta 10 millones. Con lo que se obtuvo la solucion tanto para la integral triple (volumétrica), como para la integral doble (superficial). Los resultados fueron colocados en "data frames" , gracias al uso del módulo pandas, con la finalidad de poder observar los diferentes resultados arrojados por el método montecarlo, conforme se aumentaba la cantidad de términos utilizados.

Como cabria eserar, el mejor resultado se obtuvo con 10 millones de terminos, dicho resultado fue comparado con el valor real de la carga encerrada, calculado mediante integración directa, lo que dio un error de 0.0099024 (para el caso de la última corrida llevada a cabo), el cual es un error bastante aceptable, dada la simplicidad del método Montecarlo.


# Referencias

Becerra, A., 2007​

Electrostática en presencia de dieléctricos, 2011​

Griffiths, D. J. (2005). Introduction to electrodynamics. New Jersey: Prentice Hall.​

Jackson, J. D. (2007). Classical electrodynamics. John Wiley & Sons.​

Redondo Quintera, F. & Redondo Melchor, R., 2007
